src/daemon/daemon.c

```diff
@@ -251,14 +252,18 @@ static gboolean do_exit(gpointer user_data)
 	return FALSE;
 }
 
-static void add_exit_timeout(NotifyDaemon* daemon)
+static void add_exit_timeout_if_needed(NotifyDaemon* daemon)
 {
 	g_assert (daemon != NULL);
 
-	if (daemon->exit_timeout_source > 0)
+	if (daemon->exit_timeout_source > 0 ||
+		g_hash_table_size(daemon->notification_hash) > 0 ||
+		exit_timeout_seconds < 0)
```

Comment:
Before, callers to this function checked if daemon->notification_hash was empty. Now that check is moved into this function. The new check of if the exit timeout is enabled at all is also here.